nav2_experimental/nav2_rl/nav2_turtlebot3_rl/random_crawl_action.py

```diff
@@ -0,0 +1,114 @@
+
+import time
+
+from nav2_msgs.action import RandomCrawl
+from rclpy.executors import MultiThreadedExecutor
+from rclpy.callback_groups import ReentrantCallbackGroup
+import threading
+
+class RandomCrawlActionServer(Node):
+
+    def __init__(self):
+        super().__init__('action_server')
+        
+        self.env = TurtlebotEnv()
+        self.action_size = self.env.action_space()
+        print(self.action_size)
+        self.state = self.env.reset()
+        self.observation_space = len(self.state)
+        self.state = np.reshape(self.state, [1, self.observation_space])
+        pkg_share_directory = get_package_share_directory('nav2_turtlebot3_rl')
+        path = os.path.join(pkg_share_directory,"saved_models/random_crawl_waffle.h5")
+        self.model = load_model(path)
+        self.cancel_action = False
+        q = self.model.predict(self.state)
+        
+        self._goal_handle = None
+        self._goal_lock = threading.Lock()
+        self._action_server = ActionServer(
+            self,
+            RandomCrawl,
+            'RandomCrawl',
+            execute_callback=self.execute_callback,
+            handle_accepted_callback=self.handle_accepted_callback,
+            goal_callback=self.goal_callback,
+            cancel_callback=self.cancel_callback,
+            callback_group=ReentrantCallbackGroup())
+
+    def destroy(self):
+        self.env.cleanup()
+        self._action_server.destroy()
+        super().destroy_node()
+
+    def goal_callback(self, goal_request):
+        self.get_logger().info('Received goal request')
+        self.cancel_action = False
+        return GoalResponse.ACCEPT
+    
+    def handle_accepted_callback(self, goal_handle):
+        with self._goal_lock:
+            # This server only allows one goal at a time
+            if self._goal_handle is not None and self._goal_handle.is_active:
+                self.get_logger().info('Aborting previous goal')
+                # Abort the existing goal
+                self._goal_handle.abort()
+            self._goal_handle = goal_handle
+
+        goal_handle.execute()
+
+    def cancel_callback(self, goal_handle):
+        self.get_logger().info('Received cancel request')
+        self.cancel_action = True
+        return CancelResponse.ACCEPT
+
+    def execute_callback(self, goal_handle):
+        while not self.cancel_action and rclpy.ok():
+            q_values = self.model.predict(self.state)
+            action = np.argmax(q_values)
+            next_state, reward, terminal = self.env.step(action)
+            next_state = np.reshape(next_state, [1, self.observation_space])
+            self.state = next_state
+            sleep(parameters.LOOP_RATE)
+        goal_handle.succeed()
+        self.env.stop_action()
+        result = RandomCrawl.Result()
+        return result
+
+def main(args=None):
+    rclpy.init(args=args)
+
+    action_server = RandomCrawlActionServer()
+    executor = MultiThreadedExecutor()
```

Comment:
I think you need the default `SingeThreadedExecutor`. Ising a multithreaded executor (and looking at the implementation) would cause to process multiple goals concurrently.

Reply:
@orduno @mjeronimo I tried the `SingleThreadedExecutor` first, but since I'm instantiating `TurtlebotEnv` which has it's own thread internally,  `SingleThreadedExecutor` did not work and I had to use `MultiThreadedExecutor` to make it work. Please take a look at line 62 of `turtlebot3_env.py`.

Reply:
Yes, I think we should get to the bottom of why the SingleThreadedExecutor doesn't work in this case, but I'm OK if that is filed as an issue and this PR is able to move forward.

Reply:
I don't think we need to rush to get this in. I'd prefer we understand why this workaround is necessary, before committing it. Maybe it is masking an actual bug in the code.

Reply:
IMO, the utility of sharing some experimental RL code outweighs the single vs. multi-threaded issue, since it would let people get started with this package. In the single threaded executor case, the messages weren't getting sent. In the multi-threaded executor case, they were. So, it's a communication issue (no communication versus messages flowing), not something that is affecting the computation. All that being said, if someone can root cause the issue relatively quickly, that would be great.

Reply:
@mjeronimo @orduno @crdelsey.  I and @crdelsey spent some time to debug and find the root cause.  Per our finding, we plan to keep the Multi-threaded executor to be able to have preempted functionality. However, the main issue was that in the turtlebot3_env, we were adding spin to the same executor which could cause a race condition.  To fix this issue, we decided to add a single executor in the turtlebot3_env.